src/global_timestamp_reader.cpp

```diff
@@ -0,0 +1,229 @@
+        double sum_y(0);
+        double sum_xy(0);
+        double sum_x2(0);
+        double n(static_cast<double>(_last_values.size()));
+        CSample base_sample = _last_values.back();
+        double a(1);
+        double b(0);
+        if (n > 1)
+        {
+            for (auto sample = _last_values.begin(); sample != _last_values.end(); sample++)
+            {
+                CSample crnt_sample(*sample);
+                crnt_sample -= base_sample;
+                sum_x += crnt_sample._x;
+                sum_y += crnt_sample._y;
+                sum_xy += (crnt_sample._x * crnt_sample._y);
+                sum_x2 += (crnt_sample._x * crnt_sample._x);
+            }
+            b = (sum_y*sum_x2 - sum_x * sum_xy) / (n*sum_x2 - sum_x * sum_x);
+            a = (n*sum_xy - sum_x * sum_y) / (n*sum_x2 - sum_x * sum_x);
+        }
+        std::lock_guard<std::recursive_mutex> lock(_stat_mtx);
+        LOG_DEBUG("CLinearCoefficients::calc_linear_coefs - lock");
+        _base_sample = base_sample;
+        _a = a;
+        _b = b;
+        LOG_DEBUG("CLinearCoefficients::calc_linear_coefs - unlock");
+    }
+
+    double CLinearCoefficients::calc_value(double x) const
+    {
+        LOG_DEBUG("CLinearCoefficients::calc_value - in");
+        std::lock_guard<std::recursive_mutex> lock(_stat_mtx);
+        LOG_DEBUG("CLinearCoefficients::calc_value - lock");
+        double y(_a * (x - _base_sample._x) + _b + _base_sample._y);
+        LOG_DEBUG("CLinearCoefficients::calc_value - unlock");
+        return y;
+    }
+
+    time_diff_keeper::time_diff_keeper(device* dev, const unsigned int sampling_interval_ms) :
+        _device(dev),
+        _poll_intervals_ms(sampling_interval_ms),
+        _last_sample_hw_time(1e+200),
+        _coefs(15),
+        _active_object([this](dispatcher::cancellable_timer cancellable_timer)
+            {
+                polling(cancellable_timer);
+            })
+    {
+        LOG_DEBUG("start new time_diff_keeper ");
+    }
+
+    void time_diff_keeper::start()
+    {
+        _active_object.start();
+    }
+
+    time_diff_keeper::~time_diff_keeper()
+    {
+        _active_object.stop();
+    }
+
+    bool time_diff_keeper::update_diff_time()
+    {
+        using namespace std::chrono;
+        const static double diffThresh = 500;
+        try
+        {
+            LOG_DEBUG("time_diff_keeper::update_diff_time - in");
+            std::lock_guard<std::recursive_mutex> lock(_mtx);
+            LOG_DEBUG("time_diff_keeper::update_diff_time - lock");
+            double system_time = static_cast<double>(duration_cast<nanoseconds>(system_clock::now().time_since_epoch()).count()) * TIMESTAMP_USEC_TO_MSEC * TIMESTAMP_USEC_TO_MSEC;
+            double sample_hw_time = _device->get_device_time();
+            if (sample_hw_time < _last_sample_hw_time)
+            {
+                // A time loop happend:
+                LOG_DEBUG("time_diff_keeper::call reset()");
+                _coefs.reset();
```

Comment:
What is the expected behavior when the timestamps wrap?

Reply:
When timestamp wraps, the time samples for calculating the equation for firmware time to system time are erased. Coefficients a=1, b=0 are used for the next 100 milliseconds, until some samples accumulate again. The assumption is that in that time, the firmware time and system time will not diverge.